retrofit/src/main/java/retrofit2/Retrofit.java

```diff
@@ -212,38 +212,53 @@ private void validateServiceInterface(Class<?> service) {
-        if (lookup == null) {
-          // On successful lock insertion, perform the work and update the map before releasing.
-          // Other threads may be waiting on lock now and will expect the parsed model.
-          ServiceMethod<Object> result = ServiceMethod.parseAnnotations(this, service, method);
-          serviceMethodCache.put(method, result);
-          return result;
+      if (lookup == null) {
+        // Map does not contain any value. Try to put in a lock for this method. We MUST synchronize
+        // on the lock before it is visible to others via the map to signal we are doing the work.
+        Object lock = new Object();
+        synchronized (lock) {
+          lookup = serviceMethodCache.putIfAbsent(method, lock);
+          if (lookup == null) {
+            // On successful lock insertion, perform the work and update the map before releasing.
+            // Other threads may be waiting on lock now and will expect the parsed model.
+            ServiceMethod<Object> result;
+            try {
+              result = ServiceMethod.parseAnnotations(this, service, method);
+            } catch (Throwable e) {
+              // Remove the lock on failure. Any other locked threads will retry as a result.
+              serviceMethodCache.remove(method);
+              throw e;
+            }
+            serviceMethodCache.put(method, result);
+            return result;
+          }
         }
       }
-    }
 
-    // Either the initial lookup or the attempt to put our lock in the map has returned someone
-    // else's lock. This means they are doing the parsing, and will update the map before releasing
-    // the lock. Once we can take the lock, the map is guaranteed to contain the model.
-    // Note: There's a chance that our effort to put a lock into the map has actually returned a
-    // finished model instead of a lock. In that case this code will perform a pointless lock and
-    // redundant lookup in the map of the same instance. This is rare, and ultimately harmless.
-    synchronized (lookup) {
-      return (ServiceMethod<?>) serviceMethodCache.get(method);
+      // Either the initial lookup or the attempt to put our lock in the map has returned someone
+      // else's lock. This means they are doing the parsing, and will update the map before
+      // releasing
+      // the lock. Once we can take the lock, the map is guaranteed to contain the model or null.
+      // Note: There's a chance that our effort to put a lock into the map has actually returned a
+      // finished model instead of a lock. In that case this code will perform a pointless lock and
+      // redundant lookup in the map of the same instance. This is rare, and ultimately harmless.
+      synchronized (lookup) {
+        Object result = serviceMethodCache.get(method);
+        if (result == null) {
+          // The other thread failed its parsing. We will retry (and probably also fail).
+          continue;
+        }
```

Comment:
And here (plus the enclosing `while (true)`)

Reply:
I am thinking about whether this situation is possible: `thread 1` put its `lock1`to the cache, granted the right to parse the method, but failed and threw an exception. Then, `thread 2`, which was waiting for `thread 1`  to parse the method, synchronize `lock1` successfully, find the `result` is null, continue the while loop and put its `lock2` to the cache. Then, `Thread 3` synchronize `lock1`. However, the `result` retrieved from the cache by `thread 3` is `lock2` instead of `null`. Thus, `thread 3` try to cast object to ServiceMethod and a `ClassCastException` happens. Can we check `result instanceof ServiceMethod<?>` here? What's more, can we check `lookup instanceof ServiceMethod<?>` before synchronize `lookup` to eliminate the pointless lock and redundant lookup you've metioned in comment(line 253)? Thanks.